components/nimbus/src/stateful/gecko_prefs.rs

```diff
@@ -0,0 +1,328 @@
+    /// 2. It updates the GeckoPrefStore state, such that the appropriate GeckoPrefState's
+    ///    `enrollment_value` reflects the appropriate value.
+    pub fn map_gecko_prefs_to_enrollment_slugs_and_update_store(
+        &self,
+        // contains full experiment metadata
+        experiments: &[Experiment],
+        // contains enrollment status for a given experiment
+        enrollments: &[ExperimentEnrollment],
+        // contains slug of enrolled branch
+        experiments_by_slug: &HashMap<String, EnrolledExperiment>,
+    ) -> HashMap<String, HashSet<String>> {
+        struct RecipeData<'a> {
+            experiment: &'a Experiment,
+            experiment_enrollment: &'a ExperimentEnrollment,
+            branch_slug: &'a str,
+        }
+
+        let mut state = self.get_mutable_pref_state();
+
+        /* List of tuples that contain recipe slug, rollout bool, list of feature ids, and
+         * branch, in that order.
+         */
+        let mut recipe_data: Vec<RecipeData> = vec![];
+
+        for experiment_enrollment in enrollments {
+            let experiment = match experiments
+                .iter()
+                .find(|experiment| experiment.slug == experiment_enrollment.slug)
+            {
+                Some(exp) => exp,
+                None => continue,
+            };
+            recipe_data.push(RecipeData {
+                experiment,
+                experiment_enrollment,
+                branch_slug: match experiments_by_slug.get(&experiment.slug) {
+                    Some(ee) => &ee.branch_slug,
+                    None => continue,
+                },
+            });
+        }
+        // sort `recipe_data` such that rollouts are applied before experiments
+        recipe_data.sort_by(
+            |a, b| match (a.experiment.is_rollout, b.experiment.is_rollout) {
+                (true, false) => Ordering::Less,
+                (false, true) => Ordering::Greater,
+                _ => Ordering::Equal,
+            },
+        );
+
+        /* This map will ultimately be returned from the function, as a map of pref strings to
+         * relevant enrolled recipe slugs, 'relevant' meaning experiments whose enrolled branch
+         * values apply a value to a prop for which there is a Gecko pref.
+         *
+         * We start by iterating mutably over the map of features to props to gecko prefs.
+         */
+        let mut results: HashMap<String, HashSet<String>> = HashMap::new();
+
+        for (feature_name, props) in state.gecko_prefs_with_state.iter_mut() {
+            let mut has_matching_recipes = false;
+            for RecipeData {
+                experiment:
+                    Experiment {
+                        slug,
+                        feature_ids,
+                        branches,
+                        ..
+                    },
+                experiment_enrollment,
+                branch_slug,
+            } in recipe_data.iter()
```

Comment:
```suggestion
            } in &recipe_data
```

&Vec<T> will deref-coerce to &[T]